File: tests/performance/drivers/core/app/src/put_execution.rs
//! PUT/GET execution and performance measurement

type Result<T> = std::result::Result<T, String>;
use regex::Regex;
use sf_core::protobuf::generated::database_driver_v1::*;
use std::fs;
use std::path::Path;
use std::time::{Duration, Instant};

use crate::connection::{DriverRuntime, reset_statement_query};
use crate::resource_monitor::ResourceMonitor;
use crate::results::{
    current_unix_timestamp_ms, print_statistics_put_get, write_csv_results_put_get,
    write_memory_timeline, write_metadata_if_not_replay,
};
use crate::types::PutGetResult;

pub fn execute_put_get_test(
    rt: &DriverRuntime,
    conn_handle: ConnectionHandle,
    stmt_handle: StatementHandle,
    sql_command: &str,
    warmup_iterations: usize,
    iterations: usize,
    test_name: &str,
) -> Result<()> {
    println!("\n=== Executing PUT_GET Test ===");
    println!("Query: {sql_command}");

    // Warmup
    run_warmup_put_get(rt, stmt_handle, sql_command, warmup_iterations)
        .map_err(|e| format!("Warmup phase failed: {e:?}"))?;

    if warmup_iterations > 0 {
        reset_statement_query(rt, stmt_handle, sql_command)
            .map_err(|e| format!("Failed to reset statement after warmup: {e:?}"))?;
    }

    let mut monitor = ResourceMonitor::new(Duration::from_millis(100));
    monitor.start();

    // Execute
    let results = run_test_iterations_put_get(rt, stmt_handle, sql_command, iterations)
        .map_err(|e| format!("Test phase failed: {e:?}"))?;

    let memory_timeline = monitor.stop();

    // Write & print
    let results_file = write_csv_results_put_get(&results, test_name)
        .map_err(|e| format!("Failed to write results: {e:?}"))?;

    write_memory_timeline(&memory_timeline, test_name);

    write_metadata_if_not_replay(rt, conn_handle)?;

    print_statistics_put_get(&results);
    println!(
        "  Memory timeline: {} samples collected",
        memory_timeline.len()
    );

    println!("\n✓ Complete → {results_file}");

    Ok(())
}

pub fn run_warmup_put_get(
    rt: &DriverRuntime,
    stmt_handle: StatementHandle,
    sql: &str,
    warmup_iterations: usize,
) -> Result<()> {
    if warmup_iterations == 0 {
        return Ok(());
    }

    for i in 0..warmup_iterations {
        execute_put_get_iteration(rt, stmt_handle, sql)?;

        if i < warmup_iterations - 1 {
            reset_statement_query(rt, stmt_handle, sql)?;
        }
    }
    Ok(())
}

pub fn run_test_iterations_put_get(
    rt: &DriverRuntime,
    stmt_handle: StatementHandle,
    sql: &str,
    iterations: usize,
) -> Result<Vec<PutGetResult>> {
    let mut results = Vec::with_capacity(iterations);

    for i in 0..iterations {
        let (query_time, cpu_time_s, peak_rss_mb) =
            execute_put_get_iteration(rt, stmt_handle, sql)?;

        results.push(PutGetResult {
            timestamp: current_unix_timestamp_ms(),
            query_time_s: query_time,
            cpu_time_s,
            peak_rss_mb,
        });

        if i < iterations - 1 {
            reset_statement_query(rt, stmt_handle, sql)?;
        }
    }

    Ok(results)
}

fn execute_put_get_iteration(
    rt: &DriverRuntime,
    stmt_handle: StatementHandle,
    sql: &str,
) -> Result<(f64, f64, f64)> {
    use crate::resource_monitor::{get_peak_rss_mb, process_cpu_seconds};

    create_get_target_directory(sql)?;

    let cpu_before = process_cpu_seconds();
    let start_query = Instant::now();
    rt.block_on(async |c| {
        c.statement_execute_query(StatementExecuteQueryRequest {
            stmt_handle: Some(stmt_handle),
            bindings: None,
        })
        .await
    })
    .map_err(|e| format!("PUT/GET execution failed: {e:?}"))?;
    let query_time = start_query.elapsed().as_secs_f64();
    let cpu_time_s = process_cpu_seconds() - cpu_before;
    let peak_rss_mb = get_peak_rss_mb();

    Ok((query_time, cpu_time_s, peak_rss_mb))
}

/// For GET commands:
/// - Removes existing directory to ensure clean iteration
/// - Creates fresh directory structure
fn create_get_target_directory(sql: &str) -> Result<()> {
    let sql_upper = sql.trim().to_uppercase();
    if sql_upper.starts_with("GET") || sql_upper.contains(" GET ") {
        let re = Regex::new(r"file://([^\s]+)").unwrap();
        if let Some(captures) = re.captures(sql) {
            if let Some(path_match) = captures.get(1) {
                let target_path = path_match.as_str();
                if Path::new(target_path).exists() {
                    fs::remove_dir_all(target_path).map_err(|e| {
                        format!("Failed to remove directory {}: {:?}", target_path, e)
                    })?;
                }
                fs::create_dir_all(target_path)
                    .map_err(|e| format!("Failed to create directory {}: {:?}", target_path, e))?;
            }
        }
    }
    Ok(())
}

File: tests/performance/drivers/python/app/put_execution.py
"""PUT/GET execution and performance measurement."""

import os
import re
import shutil
import time
from common import run_warmup, run_test_iterations, print_timing_stats, get_peak_rss_mb
from resource_monitor import ResourceMonitor


def execute_put_get_test(cursor, sql_command, warmup_iterations, iterations):
    """
    Execute a complete PUT/GET test: warmup, iterations, and statistics.
    
    Returns:
        tuple: (results list, memory_timeline list)
    """
    print("\n=== Executing PUT_GET Test ===")
    print(f"Query: {sql_command}")
    
    run_warmup(_execute_put_get, cursor, sql_command, warmup_iterations)

    monitor = ResourceMonitor(interval_s=0.1)
    monitor.start()

    results = run_test_iterations(_execute_put_get, cursor, sql_command, iterations)

    memory_timeline = monitor.stop()

    print_statistics(results)
    print(f"  Memory timeline: {len(memory_timeline)} samples collected")
    
    return results, memory_timeline


def print_statistics(results):
    """Print summary statistics for test results."""
    query_times = [r['query_time_s'] for r in results]
    
    print(f"\nSummary:")
    print_timing_stats("Operation time", query_times)


def _execute_put_get(cursor, sql):
    """
    Execute a PUT or GET command and collect metrics.
    
    Returns:
        dict: Dictionary with timestamp, query_time_s, cpu_time_s, and peak_rss_mb
    """
    _create_get_target_directory(sql)
    
    cpu_start = time.process_time()

    query_start = time.time()
    cursor.execute(sql)
    query_time = time.time() - query_start

    cpu_time_s = time.process_time() - cpu_start
    peak_rss_mb = get_peak_rss_mb()

    timestamp = int(time.time() * 1000)
    
    return {
        "timestamp": timestamp,
        "query_time_s": query_time,
        "cpu_time_s": cpu_time_s,
        "peak_rss_mb": peak_rss_mb,
    }


def _create_get_target_directory(sql):
    """
    Prepare target directory for GET commands.
    
    For GET commands:
    - Removes existing directory to ensure clean iteration
    - Creates fresh directory structure
    """
    if sql.strip().upper().startswith('GET'):
        match = re.search(r'file://([^\s]+)', sql)
        if match:
            target_path = match.group(1)
            if os.path.exists(target_path):
                shutil.rmtree(target_path)
            os.makedirs(target_path, exist_ok=True)
